### PR TITLE
Serialize AST

### LINKED ISSUE #2121: Vyper can't serialize ast for programs that contain decimal literals
### Version Information
* vyper Version (output of `vyper --version`): 0.2.3
* OS: linux
* Python Version (output of `python --version`): 3.8.4

### What's your issue about?
The following program cannot serialize it's AST:
```vyper
var: public(decimal)

@external
def foo(a: decimal):
    self.var += 0.1
```
Gives this error:
```bash
$ vyper -f ast DecimalLiteral.vy 
TypeError: Object of type Decimal is not JSON serializable
```

### How can it be fixed?
When serializing the `ast.Decimal` node, the `Decimal.value` attribute of the node is a `decimal.Decimal` type, which is not serializable to JSON. Instead, it should be converted to an engineering string via `decimal.Decimal.to_eng_string()`. Unsure if it's possible to convert _back_ from this JSON format, that should be considered in a possible solution.

I overrided with this fix and it did the trick:
```diff
diff --git a/vyper/ast/nodes.py b/vyper/ast/nodes.py
index 6400d721..bdd642f3 100644
--- a/vyper/ast/nodes.py
+++ b/vyper/ast/nodes.py
@@ -701,6 +701,11 @@ class Decimal(Num):
             raise InvalidLiteral("Vyper supports a maximum of ten decimal points", self)
         super().validate()
 
+    def to_dict(self):
+        d = super().to_dict()
+        d["value"] = d["value"].to_eng_string()
+        return d
+
 
 class Hex(Num):
     """
```

### PR DESCRIPTION
**Pull request by iamdefinitelyahuman**
### What I did
Ensure AST nodes as dicts are serializable.  Fixes #2121 

### How I did it
* Add unique `to_dict` and `__init__` logic on the `Decimal` and `Bytes` nodes.
* Expand the test case around this behavior to also include creating and loading a JSON string based on the AST-dict.

### How to verify it
Run tests.

### Cute Animal Picture
image

### CODE REVIEW

File: vyper/ast/nodes.py [hunk 1]
```diff
@@ -738,6 +748,16 @@ class Bytes(Constant):
...
     _translated_fields = {"s": "value"}
 
+    def __init__(self, parent: Optional["VyperNode"] = None, **kwargs: dict):
+        super().__init__(parent, **kwargs)
+        if isinstance(self.value, str):
+            self.value = self.value.encode("utf8")
+
+    def to_dict(self):
+        ast_dict = super().to_dict()
+        ast_dict["value"] = self.value.decode("utf8")
```
Comment (fubuloubu) [inline vyper/ast/nodes.py:758]: would it be better to convert to hex?

File: tests/parser/ast_utils/test_ast_dict.py [hunk 2]
```diff
@@ -78,15 +80,20 @@ def test_dict_to_ast():
...
     a: uint256 = 100
     b: int128 = -22
-    c: decimal = 3.31337
+    c: decimal = -3.3133700
     d: Bytes[11] = b"oh hai mark"
-    e: address = 0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
-    f: bool = False
+    e: Bytes[1] = 0b01010101
+    f: Bytes[88] = b"\x01\x70"
+    f: String[100] = "  baka baka   "
```
Comment (fubuloubu) [inline tests/parser/ast_utils/test_ast_dict.py:87]: `f` is used twice here

File: vyper/ast/nodes.py (hunk 1 above)
Comment (iamdefinitelyahuman) [reply vyper/ast/nodes.py:758]: I think string is better. It's simpler, it more accurately preserves the original value, and using hex would make for some weird ambiguity between the `Bytes` and `Hex` nodes.

File: tests/parser/ast_utils/test_ast_dict.py (hunk 2 above)
Comment (iamdefinitelyahuman) [reply tests/parser/ast_utils/test_ast_dict.py:87]: Oopsie.

**Pull request merged**